app/controllers/api/base_controller.rb

```diff
@@ -41,7 +47,8 @@ def authenticate_with_api_key
     params_key = request.headers["Authorization"] || ""
     hashed_key = Digest::SHA256.hexdigest(params_key)
     @api_key   = ApiKey.find_by_hashed_key(hashed_key)
-    render_unauthorized unless @api_key
+    return render_unauthorized unless @api_key
+    render_soft_deleted_api_key if @api_key.soft_deleted?
```

Comment:
I am not sure if we are testing rest of verify with otp only once, however soft delete should be tested tested for all actions where it impacts.

Reply:
Done via https://github.com/rubygems/rubygems.org/pull/2944/commits/dae1b58f18962819efd2c8f930db572970da7633